File: pine-python/pine/operators/transform_normalize.py
from __future__ import annotations

from pine.cancellation import CancellationToken
from pine.errors import OperatorException
from pine.operator import (
    AbstractOperator,
    OperatorInput,
    OperatorOutput,
    OperatorParams,
)


class TransformNormalize(AbstractOperator):
    def __init__(self):
        self._method = "min_max"

    def init(self, params: OperatorParams):
        self._method = params.get_string("method", "min_max")
        if self._method != "min_max":
            raise ValueError(
                f'transform_normalize: unsupported method "{self._method}"'
            )

    def execute(
        self, token: CancellationToken, input_: OperatorInput,
        output: OperatorOutput,
    ) -> None:
        n = input_.item_count()
        if n == 0:
            return

        field = self.item_input()[0]
        output_field = self.item_output()[0]

        vals: list[float] = []
        for i in range(n):
            try:
                vals.append(_to_double(input_.item(i, field)))
            except OperatorException as e:
                raise OperatorException(
                    f"transform_normalize: item[{i}].{field}: {e}"
                ) from e

        min_val = min(vals)
        max_val = max(vals)
        range_val = max_val - min_val

        for i in range(n):
            norm = 0.0 if range_val == 0 else (vals[i] - min_val) / range_val
            output.set_item(i, output_field, norm)


def _go_type_name(v) -> str:
    """Render a Python value's type using Go's reflection terminology so error
    messages stay byte-identical with pine-{go,cpp,java}."""
    if v is None:
        return "<nil>"
    if isinstance(v, bool):
        return "bool"
    if isinstance(v, str):
        return "string"
    if isinstance(v, int):
        return "int"
    if isinstance(v, float):
        return "float64"
    if isinstance(v, list):
        return "[]interface {}"
    if isinstance(v, dict):
        return "map[string]interface {}"
    return type(v).__name__


def _to_double(v) -> float:
    if isinstance(v, bool):
        raise OperatorException("cannot convert bool to float64")
    if isinstance(v, (int, float)):
        return float(v)
    raise OperatorException(
        f"cannot convert {_go_type_name(v)} to float64"
    )

File: pine-java/src/main/java/page/liam/pine/operators/TransformNormalize.java
package page.liam.pine.operators;

import page.liam.pine.AbstractOperator;
import page.liam.pine.CancellationToken;
import page.liam.pine.GoTypeNames;
import page.liam.pine.OperatorInput;
import page.liam.pine.OperatorOutput;
import page.liam.pine.OperatorParams;
import page.liam.pine.PineErrors;

/**
 * Operator: transform_normalize
 * Metadata contract
 *   CommonInput:  []
 *   CommonOutput: []
 *   ItemInput:    [<field>]
 *   ItemOutput:   [<output_field>]
 */
public class TransformNormalize extends AbstractOperator {
    private String method;

    @Override
    public void init(OperatorParams params) {
        method = params.getString("method", "min_max");
        if (!"min_max".equals(method)) {
            throw new IllegalArgumentException("transform_normalize: unsupported method \"" + method + "\"");
        }
    }

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) throws PineErrors.OperatorException {
        int n = input.itemCount();
        if (n == 0) return;

        String field = itemInput().get(0);
        String outputField = itemOutput().get(0);

        double[] vals = new double[n];
        for (int i = 0; i < n; i++) {
            try {
                vals[i] = toDouble(input.item(i, field));
            } catch (PineErrors.OperatorException e) {
                throw new PineErrors.OperatorException("transform_normalize: item[" + i + "]." + field + ": " + e.getMessage(), e);
            }
        }

        double min = vals[0], max = vals[0];
        for (int i = 1; i < n; i++) {
            if (vals[i] < min) min = vals[i];
            if (vals[i] > max) max = vals[i];
        }

        double range = max - min;
        for (int i = 0; i < n; i++) {
            double norm = range == 0 ? 0.0 : (vals[i] - min) / range;
            output.setItem(i, outputField, norm);
        }
    }

    private static double toDouble(Object v) throws PineErrors.OperatorException {
        if (v instanceof Number) {
            return ((Number) v).doubleValue();
        }
        // Render Java types using Go reflection terminology so error messages
        // remain byte-identical with pine-{go,cpp,python}.
        throw new PineErrors.OperatorException(
            "cannot convert " + GoTypeNames.of(v) + " to float64");
    }
}
